File: api/modules/pipeline/cards/schemas.py
"""Pydantic schemas for pipeline cards."""
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Literal, Any
from uuid import UUID

from pydantic import BaseModel, Field

CnpjBasico = Annotated[str, Field(pattern=r"^\d{8}$")]


class CardCreate(BaseModel):
    cnpj_basico: CnpjBasico
    stage_id: UUID | None = None
    display_name: str | None = Field(None, max_length=200)
    estimated_value_cents: int | None = Field(None, ge=0)
    notes: str | None = Field(None, max_length=10000)


class CardPatch(BaseModel):
    display_name: str | None = Field(None, max_length=200)
    estimated_value_cents: int | None = Field(None, ge=0)
    notes: str | None = Field(None, max_length=10000)


class CardMove(BaseModel):
    stage_id: UUID
    position: int = Field(..., ge=0)


class CardRecord(BaseModel):
    id: UUID
    pipeline_id: UUID
    stage_id: UUID
    cnpj_basico: str
    position: int
    display_name: str | None = None
    estimated_value_cents: int | None
    notes: str | None
    created_at: datetime
    updated_at: datetime


class CompanySummary(BaseModel):
    razao_social: str | None
    uf: str | None


class CardWithCompany(BaseModel):
    card: CardRecord
    company: CompanySummary


class CardInPipelineSummary(BaseModel):
    pipeline_id: UUID
    pipeline_name: str
    card_id: UUID
    stage_id: UUID
    stage_name: str


class ImportBatchRecord(BaseModel):
    id: UUID
    pipeline_id: UUID
    owner_user_id: UUID
    stage_id: UUID
    filename: str | None
    file_size_bytes: int
    content_sha256: str
    total_rows: int
    created_count: int
    skipped_count: int
    created_at: datetime


ImportRowStatus = Literal["created", "skipped"]
ImportSkipReason = Literal["invalid_cnpj_format", "cnpj_not_found", "duplicate_in_pipeline"]


class ImportRowRecord(BaseModel):
    id: int
    batch_id: UUID
    line_number: int
    raw_cnpj: str
    cnpj_basico: str | None
    display_name: str | None
    card_id: UUID | None
    status: ImportRowStatus
    reason: ImportSkipReason | None
    metadata: dict[str, Any]
    created_at: datetime

File: frontend/src/features/pipeline/schemas.ts
export interface PipelineRecord {
  id: string
  owner_user_id: string
  name: string
  description: string | null
  archived_at: string | null
  created_at: string
  updated_at: string
}

export interface StageRecord {
  id: string
  pipeline_id: string
  name: string
  position: number
  color: string | null
  is_won: boolean
  is_lost: boolean
  created_at: string
  updated_at: string
}

export interface CardRecord {
  id: string
  pipeline_id: string
  stage_id: string
  cnpj_basico: string
  position: number
  display_name: string | null
  estimated_value_cents: number | null
  notes: string | null
  created_at: string
  updated_at: string
}

export interface CompanySummary {
  razao_social: string | null
  uf: string | null
}

export interface CardWithCompany {
  card: CardRecord
  company: CompanySummary
}

export interface ImportBatchRecord {
  id: string
  pipeline_id: string
  owner_user_id: string
  stage_id: string
  filename: string | null
  file_size_bytes: number
  content_sha256: string
  total_rows: number
  created_count: number
  skipped_count: number
  created_at: string
}

export type ImportSkipReason = 'invalid_cnpj_format' | 'cnpj_not_found' | 'duplicate_in_pipeline'

export interface ImportRowRecord {
  id: number
  batch_id: string
  line_number: number
  raw_cnpj: string
  cnpj_basico: string | null
  display_name: string | null
  card_id: string | null
  status: 'created' | 'skipped'
  reason: ImportSkipReason | null
  metadata: Record<string, unknown>
  created_at: string
}

export interface ImportResult {
  batch: ImportBatchRecord
  created: number
  skipped: Array<{ line: number; cnpj: string; reason: ImportSkipReason }>
  summary: {
    total_rows: number
    valid_rows: number
    invalid_rows: number
    duplicates_in_file: number
  }
}

export type ActivityKind = 'note' | 'call' | 'email' | 'meeting'

export interface ActivityRecord {
  id: string
  card_id: string
  author_user_id: string
  kind: ActivityKind
  body: string
  occurred_at: string
  created_at: string
}

export interface TaskRecord {
  id: string
  card_id: string
  assignee_user_id: string
  title: string
  due_at: string | null
  done_at: string | null
  created_at: string
  updated_at: string
}

export interface PipelineBoardData {
  pipelines: PipelineRecord[]
  pipeline: PipelineRecord | null
  stages: StageRecord[]
  cards: CardWithCompany[]
}
